kernel/init.c

```diff
@@ -433,10 +446,9 @@ static void bg_thread_main(void *unused1, void *unused2, void *unused3)
 #endif /* CONFIG_STACK_POINTER_RANDOM */
 	boot_banner();
 
-#if defined(CONFIG_CPP)
-	void z_cpp_init_static(void);
-	z_cpp_init_static();
-#endif /* CONFIG_CPP */
+#ifdef CONFIG_STATIC_INIT_GNU
```

Comment:
This would be removing support for constructors for the arc mwdt toolchain.

Reply:
That seems bad. I'm really unclear as to how those constructors differ from the GNU constructors and whether we can support all of them in a single chunk of linker script.

Reply:
@keith-packard 

For MWDT we just call internal `__do_global_ctors_aux` function via `z_cpp_init_static`:

https://github.com/zephyrproject-rtos/zephyr/blob/e5c05da9cbda7bc68aa84b796b828767b5eea6d7/lib/cpp/abi/cpp_init.c#L27

Reply:
So MWDT places all constructors in .ctors sections, just like some GCC architectures (most GCC targets now use .init_array for "reasons", I assume). I'm not seeing any incompatibility between MWDT and GCC here; we can just collect all of the .ctors and .init_array sections together into one giant array of initialization functions and call them in the right order. That should make everything the same, eliminating any MWDT specific bits.

The current v2/linker.ld script places the .ctors sections between `_fctors` and `_ectors`, which looks like it might be specific to some arc initialization sequence which is not known to Zephyr? Does that ever get used? I can't see how it could work...

Reply:
I've pushed a separate patch in this PR which removes the special case for MWDT toolchains; I'd appreciate help understanding if this is correct or not.

Reply:
> The current v2/linker.ld script places the .ctors sections between _fctors and _ectors, which looks like it might be specific to some arc initialization sequence which is not known to Zephyr? 

I guess to whatever that toolchain provided `__do_global_ctors_aux()` does, which is not something I can see. But why not just leave the MWDT specific handling there by now? (we don't need to clean up everything in one go, specially if we risk breaking things we can't even try).

Reply:
The MetaWare `__do_global_ctors_aux` and `_fctors/_ectors` do something similar as the generic 'GNU' version, but it doesn't feel right to remove it. I'm not sure they indeed are fully equivalent, and if they are today, that may change in a next MetaWare version.
@evgeniy-paltsev, what do you think? Could we try and align/reuse GNU here, or should we stick with the MetaWare toolchain initialization?

Reply:
> But why not just leave the MWDT specific handling there by now? (we don't need to clean up everything in one go, specially if we risk breaking things we can't even try).

Yup, we figured that out yesterday :-). So, #74682 will be merged for 3.7 with just the fixes required to fix the bug.

Reply:
> The MetaWare `__do_global_ctors_aux` and `_fctors/_ectors` do something similar as the generic 'GNU' version, but it doesn't feel right to remove it. 

Unless `_fctors/_ectors` are used by code outside the Zephyr tree, then the current MWDT bits don't seem like they could possibly work? It feels like this was ported from some other system which used these symbols in the same way that Zephyr uses `__zephyr_init_array_start/__zephyr_init_array_end`.

MWDT *is* different from the generic GNU version currently though -- it uses `.ctors` sections but doesn't construct a "gnu compatible" ctors section. A GNU compatible ctors section consists of a count word followed by an array of function pointers and finally a NULL pointer.  The current MWDT ctors bits in Zephyr don't include the count or the trailing NULL.

This PR makes the GNU code match the MWDT code, removing the leading count and trailing NULL for the .ctors section.

> I'm not sure they indeed are fully equivalent, and if they are today, that may change in a next MetaWare version. @evgeniy-paltsev, what do you think? Could we try and align/reuse GNU here, or should we stick with the MetaWare toolchain initialization?

From what I can see, the MWDT compiler places all C++ constructor addresses in `.ctors` sections. If you use the Zephyr initialization code (which I think you must so that constructors can use `new`), then there isn't really any further MWDT dependency -- the Zephyr linker script defines how those addresses are laid out in memory, and then the Zephyr initialization code defines how that memory area is processed at runtime.

Reading through the ARC libgcc code leaves me with some questions though:

https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=libgcc/config/arc/initfini.c;h=0183dd495ac5db165cc706ba2d62d9431e638972;hb=refs/heads/master

In that code, destructors are run in forward order while constructors are run in *reverse* order. All other systems run constructors in forward order (and destructors in reverse order).